### ISSUE TITLE
Type check in IIFE (for JavaScript)

### ISSUE THREAD

**Issue by wehrstedt** [microsoft/TypeScript#31745]
- VSCode Version: 1.0.34 (user setup)
- OS Version: Windows (v1903)

Steps to Reproduce:

1. Use the passed code:
```js
// @ts-check
// @ts-ignore
return (function() {

	var a = 0;
	a = "1";

})();
```
No error is reported for the assignment in line 6 (the script is executed in our customized javascript runtime where global returns are allowed. Thats the reason for the `// @ts-ignore` comment).

2. Remove the return statement. You will get an error in line 6 because `a` is a number and the expression results in a string. Thats the expected behaviour:
![image](https://user-images.githubusercontent.com/27811714/58788710-be1f2900-85ec-11e9-8f43-8338710fa61d.png)

3. When you add the return statement again, the error disappears:
![image](https://user-images.githubusercontent.com/27811714/58788748-d727da00-85ec-11e9-96c9-e8e0fd689b6b.png)

I first thought that it is caused by the `@ts-ignore` comment. But removing this just results in reporting TS[1108] but not the error shown in (2):
![image](https://user-images.githubusercontent.com/27811714/58788861-17875800-85ed-11e9-974f-e7e262ede333.png)

Does this issue occur when all extensions are disabled?: Yes

**Comment by mjbvz** [comment]
As you note, this is not related to `// @ts-ignore` but to the global return, which causes TS to not analyze the function body. This seems like reasonable behavior to me since the global return means the file is syntactically invalid

**Comment by wehrstedt** [comment]
Hmm, thats understandable but in our use case its possible to do a global return. Its a server software which can be extended with  Javascript. The global return value signals e. g. the success or the failure of the script. Not the best solution but can't be changed for historical reasons.

Any chances to change this behaviour? I can not see any disadvantages supporting this. Also there are plenty of other peoples which runs Javascript in a controlled environment which supports global return values, the would probably benefit, too.

**Comment by matt-tingen** [comment]
An environment which allows global `return`s is not a JavaScript environment, but some superset thereof.

In addition to separating the `return` so it can be ignored as you pointed out above, you could potentially create a babel plugin (or similar) as a workaround. For instance, the separation and addition of `// @ts-ignore` could be automated for global `return`s before running the code through TS.

Alternatively, you could do something like the following and transform `globalReturn` _after_ TS runs:

```ts
declare function globalReturn(value: unknown): void;

globalReturn((function() {

	var a = 0;
	a = "1";

})());
```

**Comment by RyanCavanaugh** [comment]
We're not really set up to figure out what syntactically-invalid JS "should" mean. There are assumptions about what's legal and what's not baked in pretty deep and we don't have the budget to start carving them out to support these kinds of setups.

**Comment by nmain** [comment]
If you have the option to change things moving forward, you might consider migrating your global returns to `export default`, which is syntactically legal in modern javascript, and can easily be hooked up to do what you want in node.js:

```
// shared shim code used by your script runner
import(scriptfile).then(e => process.exit(e.default));
```

**Comment by wehrstedt** [comment]
> If you have the option to change things moving forward, you might consider migrating your global returns to `export default`, which is syntactically legal in modern javascript, and can easily be hooked up to do what you want in node.js:
> 
> ```
> // shared shim code used by your script runner
> import(scriptfile).then(e => process.exit(e.default));
> ```

Yes, that would be a workaround for this, but requires a lot of code changes.
Thought that this could help other peoples, too, because there is no disadvantage if vscode / typescript would analyse this function.

**Comment by nmain** [comment]
The fact that this is _syntactically_ not javascript makes it a much larger effort to support than some other extensions, and the disadvantage would be to take time away from development efforts that benefit the entire TypeScript community.

**Comment by fatcerberus** [comment]
Just throwing in my two cents: Global `return` *is* syntactically valid in CommonJS modules because Node wraps the module code in an IIFE before executing it.  I think I even remember seeing some modules in the wild that rely on this.
